# Lab 03: Descriptive Statistics & Aggregation

Your Name(s): Jose Cifuentes and Will Gruca

**Context:** We are analyzing a month of transaction data from a campus coffee shop. 

**Dataset Variables:** 
* `transaction_id`: Unique ID for the purchase
* `day_of_week`: Monday - Sunday
* `time_of_day`: Morning, Afternoon, Evening
* `item_type`: Coffee, Tea, Pastry, Sandwich, Merchandise
* `quantity`: Number of items purchased in the transaction
* `amount`: Total cost of the transaction
* `student_discount`: Boolean (True/False)


## Part 1: Descriptive Statistics

Before slicing the data into groups, we need to understand the overall properties of our dataset.



### 1. PREDICT
Think about a typical coffee shop. Explain your thought process for each of the questions below.
* Do you expect the *mean* (average) transaction amount to be higher or lower than the *median* transaction amount? Why? 
* What do you think the most expensive item type in the shop is?

We expect the mean transaction amount to be higher than the median because most customers will probably make smaller purchases, such as a coffee or like a sweet thing like a cake, while a few customers may make much larger purchases. These expensive transactions would pull the mean upward more than the median. We think merchandise will be the most expensive item type because merchandise is usually more expensive than individual food or drink items.

### 2. RUN
Execute the cell below to generate the descriptive statistics for the `amount` column.

In [1]:
# read in data
import pandas as pd
df = pd.read_csv('campus_coffee.csv')

df['amount'].describe()

count    10000.000000
mean         7.981092
std          9.058046
min          2.230000
25%          3.910000
50%          4.780000
75%          9.152500
max        213.330000
Name: amount, dtype: float64

### 3. EXPLAIN
Look at the output above. 
* Was your prediction about the mean vs. median correct? 
* The `max` amount is much higher than the 75th percentile (`75%`). In the context of our coffee shop, what real-world purchases might explain this long right tail?

My prediction was correct because the mean was $7.98, which was higher than the median of $4.78. This means that a few expensive purchases increased the average. The maximum was $213.33, which could be from someone buying merchandise and a few multiple items in that order .


## Part 2: Aggregation (Split-Apply-Combine)

Now, we want to compare different categories. Let's look at purchasing behavior based on the time of day.

### 1. PREDICT
* Which `time_of_day` (Morning, Afternoon, or Evening) do you expect to have the highest *total number of sales* (count)? 
* Which `time_of_day` do you expect to have the highest *average transaction price* (mean)? 

We would expect the highest total number of sales to be in the morning. This is due to majority of people working corporate jobs where they need to be in the office early, and need a pick me up in the morning to wake up. The highest average transaction price we think would be lunchtime/afternoon. We think this because of people going on coffee runs for the second half of their work day for their team that they work with. This would mean they are buying multiple coffees whereas in the morning it is only normally just for themselves.

### 2. RUN
Execute the following two cells to test your predictions.

In [ ]:
# Count of transactions by time of day
df.groupby('time_of_day')['transaction_id'].count()

In [ ]:
# Average price by time of day
df.groupby('time_of_day')['amount'].mean()

### 3. EXPLAIN
* Did the data match your expectations? 
* The *average price* is higher in the Afternoon than the Morning, but the *count* is lower. What does that tell you about consumer behavior at this shop? (Hint: Think about *what* people buy at 8 AM versus 1 PM).

Yes it proved our hypothesis. Morning had almost 2000 more orders than the second most, and the afternoon barely edged out the evening transaction amounts. Like I mentioned in our hypothesis, in the morning people are more likely to buy an individual coffee for themself rather then multiple. But in the afternoon, people in office jobs will go and get multiple coffees for their whole team, and this correlates to the evening, but not as much.

## Part 3: Multiple Aggregations 

Often, a single metric doesn't tell the whole story. We need to see multiple summaries side-by-side. 

### 1. PREDICT
We are going to look at the `item_type` category. 
* Which item type generates the most *total revenue* (sum)?
* Which item type is sold most *frequently* (count)?

We predict that merchandise will generate the most total revenue because it is probably more expensive than the food and drinks. I predict that coffee will be sold most frequently because coffee is usually the main product people buy at a coffee shop and also people tend to drink it a lot so it makes sense for us to be the most frequently item sold.

### 2. RUN
We can use the `.agg()` method to apply multiple mathematical functions to our groups at the same time.

In [2]:
# Provided Code: Total revenue and items sold per item type
item_summary = df.groupby('item_type').agg({
    'amount': 'sum', 
    'transaction_id': 'count'
})

print(item_summary)

               amount  transaction_id
item_type                            
Coffee       27249.59            4660
Merchandise   8827.82             164
Pastry       12677.88            1971
Sandwich     23766.40            1765
Tea           7289.23            1440


### 3. EXPLAIN
* Why does counting the transaction IDs give us a measure of item popularity?
* Look at the relationship between revenue (first column) and volume (second column). Is your most popular item also your biggest revenue driver? 

Counting the transaction IDs gives us a measure of item popularity because each transaction ID represents a purchase. An item with more transactions IDs appears in more transactions and is most popular. Coffee is both the most popular item and the biggest in revenue. Coffee appeared in 4,660 transactions and generated $27,249.59 in revenue, which were the highest values of all the item types. One interesting result is that merchandise had only 164 transactions but still generated $8,827.82 in revenue. This suggests that merchandise is purchased much less frequently but is much more expensive than all the other items by a lot more.

## Part 4: Temporal Analysis

Temporal analysis focuses on understanding how data changes over time.

### 1. PREDICT
We are going to look at the `day_of_week` variable. 
* Do you expect certain days of the week to have higher sales than others? Why?

I expect weekdays to have higher sales than weekends because more students are on campus attending classes during the week. I would especially expect the middle of the week to have higher sales because students are following their normal class schedules and may stop at the coffee shop between classes to do homework, socialize with friends or wait during there classes.


### 2. RUN


In [3]:
df.groupby('day_of_week').agg({'amount': ['sum', 'count']})

amount      
                  sum count
day_of_week                
Friday       10978.33  1441
Monday       14638.41  1840
Saturday      5715.26   666
Sunday        4667.52   589
Thursday     14877.15  1813
Tuesday      14543.12  1819
Wednesday    14391.13  1832

### 3. EXPLAIN
* What patterns do you notice in sales across different days of the week? 
* What real-world factors not captured in the data might influence this pattern?

Sales are highest Monday through Thursday, with around 1,800 transactions each day. They start to drop on Friday and are much lower on Saturday and Sunday, with Sunday having only 589 transactions. This could be because students are busier on campus during the week because of classes. Class schedules, campus events, and the coffee shop's weekend hours could also affect this pattern.


## Part 5: Relationships Between Variables (Covariance)

While aggregating tells us about single columns, **covariance** tells us about the *directional relationship* between two numeric variables. Do they tend to move in the same direction, opposite directions, or is there no pattern at all?

### 1. PREDICT
* Do you expect the covariance between `quantity` and total `amount` to be positive, negative, or close to zero? Why?

We expect the covariance would be positive because usually the more items you buy, the more you will have to pay. (Or the amount is higher with a higher quantity).

### 2. RUN
Execute the cell below to calculate the covariance matrix for our two numeric columns.

In [4]:
# Calculate covariance between quantity and total price
df[['quantity', 'amount']].cov()

# The result is a covariance matrix showing the relationship between `quantity` and `amount`. 
# Where each entry represents the covariance between the corresponding pair of variables. 
# The diagonal entries represent the variance of each variable.

,quantity,amount
quantity,0.662372,3.548450
amount,3.548450,82.048197


### 3. EXPLAIN
* Look at the output matrix. The covariance between `quantity` and `amount` is a positive number. What does this confirm about the relationship between the number of items purchased and total cost?
* The resulting number itself might look arbitrary. Unlike *correlation*, covariance is not on a standardized scale of -1 to 1; its raw size depends entirely on the units of the variables (items and dollars). Given this, why might data scientists prefer to present correlation rather than covariance when explaining data relationships to others?
* Using the numbers in the table above, calculate the correlation coefficient between `quantity` and `amount`. 

That confirms that the covariance is positive which means as the quantity increases the amount also increases. Scientists would likely prefer correlation because it uses that scale of -1 to 1 which makes it easier to compare as well as understand the strength and the direction of the certain relationships between the variables. The correlation coefficient is 0.48 which shows a moderate positive relaitonship between quantity and amount.

### 4. A FINAL QUESTION
* Let's assume our dataset included a variable for the daily outside temperature and that our coffee shop only sales hot coffee. Would you expect the covariance between `temperature` and the total sales of hot coffee to be positive or negative? Why?

We would expect that there would be a negative correlation becuase when the outside temperatures increase the amount of sales of hot coffee would decrease as it would be super hot out already.